### ISSUE TITLE
Incorrect suggestion of `string_lit_as_bytes` lint

### ISSUE THREAD

**Issue by ordovicia** [rust-lang/rust-clippy#4494]
Clippy version: `clippy 0.0.212 (e3cb40e 2019-06-25)`

Clippy suggests using a byte string literal instead of `"string literal".as_bytes()`.
The suggestion does not compile in the following case because the type of byte string literal `&[u8; N]` does not implement `std::io::Read`.

```rust
fn f(_: impl std::io::Read) {
}

fn main() {
    f("string literal".as_bytes());
//    ^^^^^^^^^^^^^ help: consider using a byte string literal instead: `b"string literal"`

    // f(b"string literal");    // does not compile
    // ^ the trait `std::io::Read` is not implemented for `&[u8; 14]`
    
    f(&b"string literal"[..]);  // compiles
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c943579be3ef133470f512b1cef39ba9)

Clippy should allow the above case or suggest `&b"string literal"[..]`.

**Comment by mikerite** [comment]
We should always suggest `&b"string literal"[..]`. Changing the type of the expression is asking for trouble and `&b"string literal"[..]` -> `b"string literal"` could be a separate lint anyway.

**Comment by mahkoh** [comment]
The lint should instead be removed. The only justification for the lint given is that using byte strings is shorter. `&b"string literal"[..]` is perl-tier.
